resources/top_level_overview.markdown

```diff
@@ -0,0 +1,283 @@
+set of _output_ `(key, value)` pairs.
+
+2. __Map__: A user defined function that takes input pairs and produces a set of _intermediate_ `(key, value)` pairs. The MapReduce library groups all of these
+intermediate values associated with the same intermediate key, _I_, and passes
+them to the reduce function.
+
+3. __Reduce__:  Is a user defined function that accepts an intermediate key,
+`I`, and a set of values for that key. It merges the values together.  Typically
+zero or one output values are produced by any _Reduced_ invocation.
+(The list of values are passed by an iterator to deal with memory availability.)
+
+The above computational process has the following type associations:
+
+```
+map:    (k1, v1)       -> list(k2, v2)
+
+reduce: (k2, list(v2)) -> list(v2)
+```
+
+That is, a map task, which is a copy of a user-defined function, accepts
+key-value pairs, `(k1, v1)`, and outputs `(k2, v2)`, representative of the
+intermediate value pairs, `I`.  These intermediate pairs are not necessarily of
+the same type as the input pairs.  The reducer then accepts all keys, `k2`, and
+the corresponding list of values, `list(v2)`.  The reducer then performs some
+user-defined operation on the list, and produces some output.
+
+
+#### Steps in a MapReduce Process:
+
+The following is a simplification of the steps outlined in
+_Map Reduce: Simplified Data Processing on Large Clusters_.  After we break down
+the process, we'll explore how Hadoop handles each of the following:
+
+1. The MapReduce library splits the input into _M_ chunks, and starts up copies
+of the program on multiple machines.
+
+2. The _master_ fork assigns tasks to the _worker_ nodes who are just running
+copies of the _user program_.  There are `M` map tasks, and `R` reduce tasks
+that the _master_ must assign. Note that a worker node takes on the role of
+either map, or reduce. Typically, there are many more map nodes than reduce
+nodes.
+
+3. A _worker_ assigned a map task, reads in the contents of the corresponding
+input split.  The _worker_ parses (key, values) pairs and passes it to the map
+function. The _intermediate_ (key, value) pairs produced by the mapping are
+buffered into memory.
+
+4. The local disk it partitioned into `R` regions by the _partitioning
+function_. The buffered _intermediate_ pairs produced by a map worker are
+periodically written to these partitions, with the particular location of these
+writes sent back to `master` for use by the reduce workers.
+
+5. When the `master` notifies a _reduce_ worker about a particular location of
+_intermediate_ key-value pairs, it reads the buffered data from the local
+disks of the map workers. When a reduce worker has read all intermediate data,
+it sorts it by key so that all of the same keys are grouped together. The
+sorting is needed because many different keys can map to the same reduce task.
+If the amount of intermediate data is too large to fit in memory, an external
+sort is used.
```

Comment:
let's use "out of core" instead of "external".  or "on disk".

Reply:
Sounds good, I think I'm partial to "on disk", as it keeps more with the prose of the writing.